File: Coding Practice/codingpractice/src/main/python/interviewbit/letter_phone.py
class Solution:
    # @param A : string
    # @return a list of strings
    def letterCombinations(self, A):
        result = list()
        # input = [[1], ['a', 'b', 'c'], ['d', 'e', 'f']]
        input = self.map_digit_to_letters(A)
        self.helper_method(result, [], input)
        return result

    def map_digit_to_letters(self, A):
        keypad = {
            '0': '0',
            '1': '1',
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        input_2d = list()
        for letter in A:
            val = [char for char in keypad[letter]]
            input_2d.append(val)

        return input_2d

    def helper_method(self, result, curr, rest):
        if(len(rest) == 0):
            tmp_result = ''
            for letter in curr:
                tmp_result += letter
            result.append(tmp_result)
            return result

        sublist = rest.pop(0)
        for ele in sublist:
            self.helper_method(result, curr + [ele], list(rest))


if __name__ == '__main__':
    s = Solution()
    res = s.letterCombinations('2')
    print(res)

File: Coding Practice/codingpractice/src/main/java/com/interviewbit/backtracking/LetterPhone.java
package com.interviewbit.backtracking;

import java.util.ArrayList;
import java.util.Map;

public class LetterPhone {

//    Input: Digit string "23"
//    Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].

    public static void main(String[] args) {

    LetterPhone l = new LetterPhone();
    ArrayList<String> res = l.letterCombinations("123");
    System.out.println(res);

    }

    public ArrayList<String> letterCombinations(String A) {

        Map<String, String> keypad = Map.of(
                "0", "0",
                "1", "1",
                "2", "abc",
                "3", "def",
                "4", "ghi",
                "5", "jkl",
                "6", "mno",
                "7", "pqrs",
                "8", "tuv",
                "9", "wxyz"
        );
        ArrayList<char[]> rem = new ArrayList<>();
        for (char key: A.toCharArray()) {
            String val = keypad.get(""+key);
            System.out.println("key and val are "+ key + val);
            char[] arr_val = val.toCharArray();
            rem.add(arr_val);
        }
        ArrayList<String> res = new ArrayList<>();
        ArrayList<String> curr = new ArrayList<>();

        return helper(res, curr, rem);
    }

    public static ArrayList<String> helper(ArrayList<String> res, ArrayList<String> curr, ArrayList<char[]> rem){
        if( rem.size() == 0){
            String tmp = "";
            for (int i = 0; i < curr.size(); i++) {
                tmp += curr.get(i);
            }
            res.add(tmp);
            return res;
        }

        char[] sublist = rem.remove(0);
        for (int i = 0; i < sublist.length; i++) {
            ArrayList<String> new_curr = (ArrayList<String>) curr.clone();
            new_curr.add("" + sublist[i]);
            ArrayList<char[]> new_rem = (ArrayList<char[]>) rem.clone();
            helper(res, new_curr, new_rem);
        }
        return res;
    }
}
